### DISCUSSION TITLE
How can I pause UI updates? I want to manually control UI updates temporarily

### DISCUSSION THREAD

**Discussion by Seikaijyu** [AllenDang/giu#567] [Q&A] [answered]
-

**Comment by gucio321** [comment, marked as answer]
hmm, I don't think it is possible in current implementation, why do you need it? Could you describe your use-case? Maybe you just need to call giu.Update() to trigger UI update?
Also, if you stop updating UI, nothing will not work there (buttons, text fields e.t.c.)

**Reply by Seikaijyu** [reply]
I hope to have a stable real-time frame number, as the game does, for playing component animation

**Reply by Seikaijyu** [reply]
Frame generation time is critical for animation

**Reply by Seikaijyu** [reply]
If you just call Update, you can't get a stable and smooth frame generation

**Reply by Seikaijyu** [reply]
I am currently developing a giu based framework as a subproject of the project I mentioned a few months ago: D

**Reply by gucio321** [reply]
so, for animations, you need to add your own time system, like:
```go
for <- range t.Tick(<time delta>) {
   <skip to next frame>
   giu.Update()
}
```

`<skip to next frame>` should update some variable with layout that should be rendered or probably change some index.
I think I did something similar in my gif player `examples/gifdecode`

**Reply by Seikaijyu** [reply]
Will this not lead to waste of resources? For example, the same frame is updated multiple times

**Reply by Seikaijyu** [reply]
Just now I saw a function that might fulfill my idea
```go
g.Context.GetPlatform().SetTPS(0)
```
But there is a problem with this. The value cannot be 0. Maybe we can start from here?

**Reply by Seikaijyu** [reply]
![image](https://user-images.githubusercontent.com/68497295/195609090-6ac73abf-1719-484d-aaf5-ab2ec2b9bcd4.png)
For example, make some changes

**Reply by Seikaijyu** [reply]
Well, it seems that this is not the same as FPS. It's a pity

**Reply by Seikaijyu** [reply]
If there is no other solution, I will follow your suggestion

**Reply by gucio321** [reply]
I'm trying to create FPS counter app

**Reply by gucio321** [reply]
```go
package main

import (
	"fmt"
	"time"

	"github.com/AllenDang/giu"
)

var (
	now        time.Time
	timeDelta  time.Duration
	fpsTime    time.Time
	frames     int
	currentFPS int
)

func loop() {
	giu.SingleWindow().Layout(
		giu.Custom(func() {
			frames++
			timeDelta = time.Now().Sub(now)
			now = time.Now()
			fpsTime = fpsTime.Add(timeDelta)
			if fpsTime.Second() >= 1 {
				currentFPS = frames
				frames = 0
				fpsTime = time.Time{}
			}
		}),
		giu.Labelf("Current time delta %v", timeDelta),
		giu.Labelf("Current FPS: %d", currentFPS),
	)
}

func main() {
	wnd := giu.NewMasterWindow("FPS calculation", 640, 480, 0)
	fmt.Println(giu.Context.GetPlatform().GetTPS())
	wnd.Run(loop)
}
```
ok, thats it

**Reply by Seikaijyu** [reply]
It seems that resources will not be wasted. Thank you for your answer

**Reply by gucio321** [reply]
yah, IDK why, maybe the power saving mode does something about this FPS, but calling giu.Update thousands time per second doesn't change fps :smile:
commiting my example to master ;-)
(below the code with 100 Update calls per second - let me know if I missunderstood what FPS is)
```golang
package main

import (
	"time"

	"github.com/AllenDang/giu"
)

var (
	now        time.Time
	timeDelta  time.Duration
	fpsTime    time.Time
	frames     int
	currentFPS int
)

func loop() {
	giu.SingleWindow().Layout(
		giu.Custom(func() {
			frames++
			timeDelta = time.Now().Sub(now)
			now = time.Now()
			fpsTime = fpsTime.Add(timeDelta)
			if fpsTime.Second() >= 1 {
				currentFPS = frames
				frames = 0
				fpsTime = time.Time{}
			}
		}),
		giu.Labelf("Current time delta %v", timeDelta),
		giu.Labelf("Current FPS: %d", currentFPS),
	)
}

func main() {
	wnd := giu.NewMasterWindow("FPS calculation", 640, 480, 0)
	go func() {
		for range time.Tick(time.Millisecond) {
			giu.Update()
		}
	}()
	wnd.Run(loop)
}
```

**Reply by Seikaijyu** [reply]
I suddenly found a problem. This can only ensure that the frame rate is above the set value, but it does not guarantee that the frame generation is stable at the set value. Is there any way to limit FPS?

**Reply by gucio321** [reply]
maybe try with SetTPS?

**Reply by Seikaijyu** [reply]
Using SetTPS will affect the processing speed of all events, such as OnClick OnHover. I want to have an option that only limits the number of frames

**Reply by Seikaijyu** [reply]
Can this function be implemented at present?

**Reply by AllenDang** [reply]
@SuperLiveYou May i ask whats your use case?

**Reply by Seikaijyu** [reply]
I think I don't need to think about this. Maybe it's OK to follow TPS @AllenDang
